Assets/MRTK/SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs

```diff
@@ -114,54 +157,168 @@ protected override bool IsValidController(IMixedRealityController controller)
-                        else
-                        {
-                            RotationBehavior = SolverRotationBehavior.LookAtMainCamera;
-                            OffsetBehavior = SolverOffsetBehavior.LookAtCameraRotation;
+                            return IsUserGazingAtActivationPoint(jointedHand);
                         }
                     }
+
+                    eyeGazeActivationAlreadyTriggered = palmFacingThresholdMet ? eyeGazeActivationAlreadyTriggered : false;
+
+                    return palmFacingThresholdMet;
                 }
                 else
                 {
                     Debug.LogError("HandConstraintPalmUp requires controllers of type IMixedRealityHand to perform hand activation tests.");
                 }
 
+                eyeGazeActivationAlreadyTriggered = palmFacingThresholdMet ? eyeGazeActivationAlreadyTriggered : false;
                 return palmFacingThresholdMet;
             }
 
             return true;
         }
+
+        /// <summary>
+        /// Checks to see if the palm is currently facing the user
+        /// </summary>
+        /// <param name="jointedHand"></param>
+        /// <param name="palmPose"></param>
+        /// <param name="palmCameraAngle"></param>
+        /// <returns></returns>
+        private bool IsPalmFacingUpwards(IMixedRealityHand jointedHand, MixedRealityPose palmPose, float palmCameraAngle)
+        {
+            if (requireFlatHand)
+            {
+                // Check if the triangle's normal formed from the palm, to index, to ring finger tip roughly matches the palm normal.
+                MixedRealityPose indexTipPose, ringTipPose;
+
+                if (jointedHand.TryGetJoint(TrackedHandJoint.IndexTip, out indexTipPose) &&
+                    jointedHand.TryGetJoint(TrackedHandJoint.RingTip, out ringTipPose))
+                {
+                    var handNormal = Vector3.Cross(indexTipPose.Position - palmPose.Position,
+                                                   ringTipPose.Position - indexTipPose.Position).normalized;
+                    handNormal *= (jointedHand.ControllerHandedness == Handedness.Right) ? 1.0f : -1.0f;
+
+                    if (Vector3.Angle(palmPose.Up, handNormal) > flatHandThreshold)
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            // Check if the palm angle meets the prescribed threshold
+            return palmCameraAngle < facingCameraTrackingThreshold;
+        }
+
+        /// <summary>
+        /// Checks to see if the user is currently gazing at the activation point
+        /// </summary>
+        /// <param name="jointedHand"></param>
+        /// <returns></returns>
+        private bool IsUserGazingAtActivationPoint(IMixedRealityHand jointedHand)
+        {
+            var gazeProvider = (GazeProvider)CoreServices.InputSystem.EyeGazeProvider;
```

Comment:
Would recommend using  InputRayUtils.TryGetEyeGazeRay instead, like so:

```
Ray inputRay; 
if (InputRayUtils.TryGetRay(InputSourceType.Eye, Handedness.any, out ray) || InputRayUtils.TryGetRay(InputSourceType.Head, Handedness.any, out ray))
{ 
... process and do stuff
```


That code assumes that if the eye data is not provided, it will use the head gaze.

Reply:
Done and tested!